Add/35129 product mvp ces

louwie17 opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Add's a CES bar to the footer of the Product MVP page when a user saves or publishes a product for the first time.

Closes #35129 

https://user-images.githubusercontent.com/2240960/203027445-7963d073-37a6-45ee-9e11-7a54678cd03c.mp4

<!-- The next section is mandatory. If your PR doesn't require testing, please indicate that you are purposefully omitting instructions. -->

- [ ] This PR is a very minor change/addition and does not require testing instructions (if checked you can ignore/remove the next section).

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Otherwise, please include detailed instructions on how these changes can be tested (including pre-conditions, configuration, steps to take and expected results). It may help to write your instructions using pseudocode -- as if you're telling a computer how to execute the test. -->

1. Load this branch, build it, and make sure you have the `new-product-management-experience` feature flag enabled (you can do so using the latest version of the Beta tester within the mono repo).
2. Make sure you have tracks enabled and also outputted in the console: localStorage.setItem( 'debug', 'wc-admin:*' );
3. Go to Products > Add New (MVP)
4. Add a name and click Save draft
5. A bar in the footer should show up saying: `You're using the new product editor (currently in development). How is your experience so far?` and give the option to share feedback
6. Refresh the page, the bar should show up again after the product has been loaded.
7. When clicking share feedback the CES modal should show up. You can fill it out and click Send
8. This should trigger a track event with `new_product` action and the score.
9. The bar should stay put after submitting feedback
10. Now click Publish and enter feedback again, this should trigger the `new_product` action once finished.
11. The feedback footer should still show up after, even after refresh.
12. Now click the x or `Turn it off` on the notice, the notice should not show up there after, also on refresh.

If you wanted to re-test you can delete the `woocommerce_ces_product_mvp_ces_action` option.

<!-- End testing instructions -->

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

louwie17 committed:
Add product mvp ces footer

louwie17 committed:
Add changelog

louwie17 committed:
Reset action option name after CES action

louwie17 committed:
Make sure we make use of cesAction

louwie17 committed:
Add close button

louwie17 committed:
Fix formatting

louwie17 committed:
Fix css styling issues

louwie17 committed:
Add icon to success notice

louwie17 committed:
Add product mvp CES hook and modify logic a little bit

louwie17 commented on plugins/woocommerce-admin/client/layout/footer/footer.tsx:
```diff
@@ -0,0 +1,24 @@
+import './footer.scss';
+import { WC_FOOTER_SLOT_NAME, WooFooterItem } from './utils';
+
+export const Footer: React.FC = () => {
```
A simple Footer component with a SlotFill, this allows us to easily fill it with the CES footer and correctly use the full width.

joelclimbsthings replied:
💯

louwie17 committed:
Update the CES product action name

joelclimbsthings commented on plugins/woocommerce-admin/client/layout/footer/utils.tsx:
```diff
@@ -0,0 +1,46 @@
+
+export const WC_FOOTER_SLOT_NAME = 'woocommerce_footer_item';
+/**
+ * Create a Fill for extensions to add items to the WooCommerce Admin header.
```
Typo here:
```suggestion
 * Create a Fill for extensions to add items to the WooCommerce Admin footer.
```

joelclimbsthings commented on plugins/woocommerce-admin/client/customer-effort-score-tracks/product-mvp-ces-footer.tsx:
```diff
@@ -0,0 +1,154 @@
+		cesShownForActions,
+		resolving: isLoading,
+	} = useSelect( ( select ) => {
+		const { getOption, isResolving } = select( OPTIONS_STORE_NAME );
```
Just a note here that `isResolving` will actually be `false` before it's `true`, which can cause some erratic behavior. I think it's more reliable to use `hasFinishedResolution` if that can serve your use case.

joelclimbsthings reviewed (changes requested):
Nice work @louwie17 ! Made a couple minor comments.

I also noticed that the display breaks around a viewport width of ~1050px, which seems a bit fragile even if we're not putting attention towards mobile on this screen so far. 

![image](https://user-images.githubusercontent.com/444632/203664416-e8a5d9d3-38ea-4cc8-9303-409b2bc8bc9a.png)

Once we do support mobile this will be additional technical debt, so perhaps it's worth trying to accommodate it here in some fashion? @jarekmorawski 

Also, not at all required but I wonder if there is an argument for adding the UI layer of this component to the `customer-effort-score` package?

louwie17 committed:
Address PR feedback and fix styling for smaller screens

louwie17 commented:
Thanks for the review @joelclimbsthings, I addressed your feedback and fixed the styling around the viewport.

> Also, not at all required but I wonder if there is an argument for adding the UI layer of this component to the `customer-effort-score` package?

Yeah potentially, given this is mostly geared towards the Product MVP, I don't think it be useful for users to re-use this, unless they wanted to gather feedback and have access to our `tracks` :)

The one potential item that probably should be moved is the Footer Slot fill, we did have to move that along with the header slot fill. Given the header hasn't moved yet, I will leave that until we move both, maybe when we implement extensibility.

joelclimbsthings commented on plugins/woocommerce-admin/client/customer-effort-score-tracks/product-mvp-ces-footer.tsx:
```diff
@@ -0,0 +1,167 @@
+				</WooFooterItem>
+			) }
+			{ showFeedbackModal && (
+				<CustomerFeedbackModal
```
Since I just merged the [CES: Add additional question](https://github.com/woocommerce/woocommerce/pull/35680) PR, we'll want to add the additional props here and double-check that it's working properly after a rebase on trunk. [Here's an example.](https://github.com/woocommerce/woocommerce/blob/28d7ca306bc761bb3b8a2db57f2d59451f977d80/packages/js/customer-effort-score/src/customer-effort-score.tsx#L96)

louwie17 replied:
Addressed here: https://github.com/woocommerce/woocommerce/pull/35652/commits/66ad4d08988dbf1e63d8acfd5812dbaf32bf51e0

joelclimbsthings reviewed (changes requested):
Right on, great work @louwie17 . Everything looks good, although I recently merged a PR that will require a minor update. See [this comment ](https://github.com/woocommerce/woocommerce/pull/35652#discussion_r1035242065)for details.

louwie17 committed:
Add second question to CES product footer and update options

louwie17 committed:
Add changelog

louwie17 commented:
> Right on, great work @louwie17 . Everything looks good, although I recently merged a PR that will require a minor update. See [this comment ](https://github.com/woocommerce/woocommerce/pull/35652#discussion_r1035242065)for details.

Thanks @joelclimbsthings, I updated it to account for the second question. Also note that I updated the `options` text in the customer effort modal to match the questions, as from the design that looks to be new normal. Otherwise the options sounded kinda weird.

louwie17 committed:
Fix lint error

louwie17 committed:
Fix tests

joelclimbsthings commented on packages/js/customer-effort-score/src/customer-feedback-modal/index.tsx:
```diff
@@ -49,29 +51,33 @@ function CustomerFeedbackModal( {
+		customOptions && customOptions.length > 0
+			? customOptions
+			: [
+					{
```
Nice! Thanks for updating these @louwie17

joelclimbsthings reviewed (approved):
>  Also note that I updated the options text in the customer effort modal to match the questions, as from the design that looks to be new normal. Otherwise the options sounded kinda weird.

💯 Looks like that was a miss on my part. Thanks for getting that. Everything looks and works great! 🚢
